Python/pytime.c

```diff
@@ -781,35 +869,71 @@ _PyTime_GetMonotonicClockWithInfo(_PyTime_t *tp, _Py_clock_info_t *info)
 
 
 #ifdef MS_WINDOWS
+static int
+init_win_perf_counter(LONGLONG *t0, LONGLONG *perf_frequency,
+                      _PyTime_t *num_p, _PyTime_t *den_p)
+{
+    LARGE_INTEGER now;
+    LARGE_INTEGER freq;
+    _PyTime_t num, den, gcd;
+
+    QueryPerformanceCounter(&now);
+    if (!QueryPerformanceFrequency(&freq)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
+    if (freq.QuadPart < 1) {
+        return -1;
+    }
+
+    *t0 = now.QuadPart;
+    *perf_frequency = freq.QuadPart;
+
+    /* Compute the greatest common divisor to minimize numerator and
+       denominator, to reduce the risk of integer overflow in
+       _PyTime_GetWinPerfCounterWithInfo(). */
+    num = SEC_TO_NS;
+    den = freq.QuadPart;
+    gcd = _PyTime_GCD(num, den);
+    num /= gcd;
+    den /= gcd;
+    if (num < 1 || den < 1) {
+        return -1;
+    }
+
+    *num_p = num;
+    *den_p = den;
+    return 0;
+}
+
 int
-_PyTime_GetWinPerfCounterWithInfo(_PyTime_t *t, _Py_clock_info_t *info)
+_PyTime_GetWinPerfCounterWithInfo(_PyTime_t *tp, _Py_clock_info_t *info)
 {
-    static LONGLONG cpu_frequency = 0;
-    static LONGLONG ctrStart;
+    static LONGLONG perf_frequency = 0;
+    static _PyTime_t numer = 0, denom = 0;
+    static LONGLONG t0 = 0;
     LARGE_INTEGER now;
-    double diff;
+    _PyTime_t t;
 
-    if (cpu_frequency == 0) {
-        LARGE_INTEGER freq;
-        QueryPerformanceCounter(&now);
-        ctrStart = now.QuadPart;
-        if (!QueryPerformanceFrequency(&freq) || freq.QuadPart == 0) {
-            PyErr_SetFromWindowsErr(0);
+    if (perf_frequency == 0) {
+        if (init_win_perf_counter(&t0, &perf_frequency, &numer, &denom) < 0) {
             return -1;
         }
-        cpu_frequency = freq.QuadPart;
     }
-    QueryPerformanceCounter(&now);
-    diff = (double)(now.QuadPart - ctrStart);
     if (info) {
         info->implementation = "QueryPerformanceCounter()";
-        info->resolution = 1.0 / (double)cpu_frequency;
+        info->resolution = 1.0 / (double)perf_frequency;
         info->monotonic = 1;
         info->adjustable = 0;
     }
 
-    diff = diff / (double)cpu_frequency;
-    return _PyTime_FromDouble(t, diff, _PyTime_ROUND_FLOOR, SEC_TO_NS);
+    QueryPerformanceCounter(&now);
+    t = (now.QuadPart - t0);
```

Comment:
Can this overflow?

Reply:
(now.QuadPart - t0) shouldn't overflow. Both values are supposed to be positive and the clock is *supposed* to be monotonic. In practice, there is no warranty, the clock might go backward, so it isn't used for time.monotonic() but only time.perf_counter().

Note: my PR doesn't introduced this code, it exists since Python 2.7 at least:
https://github.com/python/cpython/blob/2.7/Modules/timemodule.c#L184

Reply:
I don't mean that the subtraction can overflow. I mean that casting the result from LARGE_INTEGER to _PyTime_t can overflow. In the current code the result is casted to double. It never overflows.

Reply:
(Stupid GitHub, my rebase dropped the commits, I don't have the context anymore.)

No, in practice, _PyTime_t == LONGLONG, it's the same type. I only added an assertion to be "future proof".